File: app/src/main/java/uk/co/victoriajanedavis/chatapp/domain/interactors/IsUserLoggedIn.java
package uk.co.victoriajanedavis.chatapp.domain.interactors;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.annotations.NonNull;
import io.reactivex.annotations.Nullable;
import uk.co.victoriajanedavis.chatapp.data.repositories.TokenRepository;
import uk.co.victoriajanedavis.chatapp.domain.entities.TokenEntity;
import uk.co.victoriajanedavis.chatapp.domain.entities.TokenEntityHolder;

public class IsUserLoggedIn implements ReactiveInteractor.RetrieveInteractor<Void, Boolean> {

    @NonNull private final TokenRepository repository;
    @NonNull private final TokenEntityHolder holder;


    @Inject
    public IsUserLoggedIn(@NonNull final TokenRepository repository,
                          @NonNull final TokenEntityHolder holder) {
        this.repository = repository;
        this.holder = holder;
    }

    @NonNull
    @Override
    public Observable<Boolean> getBehaviorStream(@Nullable Void aVoid) {
        return repository.getTokenStream()
                .doOnNext(holder::setTokenEntity) // TODO: I don't love this way of setting the TokenHolder
                .map(token -> !token.isEmpty());
    }
}

File: app/src/main/java/uk/co/victoriajanedavis/chatapp/domain/interactors/IsUserLoggedIn.kt
package uk.co.victoriajanedavis.chatapp.domain.interactors

import android.util.Log
import javax.inject.Inject

import io.reactivex.Observable
import uk.co.victoriajanedavis.chatapp.data.repositories.TokenRepository
import uk.co.victoriajanedavis.chatapp.domain.entities.TokenEntityHolder
import uk.co.victoriajanedavis.chatapp.injection.scopes.ApplicationScope

@ApplicationScope
class IsUserLoggedIn @Inject constructor(
    private val repository: TokenRepository,
    private val holder: TokenEntityHolder
) : ReactiveInteractor.RetrieveInteractor<Void, Boolean> {

    override fun getBehaviorStream(params: Void?): Observable<Boolean> {
        return repository.tokenStream()
            .doOnNext { token -> holder.tokenEntity = token } // TODO: I don't love this way of setting the TokenHolder
            .map { token -> !token.isEmpty }
    }
}
